modules/compliance-tests/src/smithy4s/compliancetests/ClientHttpComplianceTestCase.scala

```diff
@@ -0,0 +1,181 @@
+
+  private def matchRequest(
+      request: Request[IO],
+      testCase: HttpRequestTestCase
+  ): IO[ComplianceResult] = {
+    val bodyAssert = testCase.body
+      .map { expectedBody =>
+        request.bodyText.compile.string.map { responseBody =>
+          assert.eql(expectedBody, responseBody)
+        }
+      }
+      .getOrElse(assert.success.pure[IO])
+
+    val headerAssert =
+      testCase.headers
+        .map { expectedHeaders =>
+          expectedHeaders
+            .map { case (name, value) =>
+              val actual = request.headers.get(CIString(name))
+              actual
+                .map { values =>
+                  val actualValue = values.map(_.value).fold
+                  assert.eql(value, actualValue)
+                }
+                .getOrElse(
+                  assert.fail(s"Header $name was not found in the response.")
+                )
+            }
+            .toList
+            .combineAll
+        }
+        .getOrElse(assert.success)
+
+    val expectedUri = baseUri
+      .withPath(
+        Uri.Path.unsafeFromString(testCase.uri)
+      )
+      .withQueryParams(
+        testCase.queryParams.combineAll.map {
+          _.split("=", 2) match {
+            case Array(k, v) =>
+              (
+                k,
+                Uri.decode(
+                  toDecode = v,
+                  charset = StandardCharsets.UTF_8,
+                  plusIsSpace = true
+                )
+              )
+          }
+        }.toMap
+      )
+
+    val uriAssert = assert.eql(expectedUri, request.uri)
+    val methodAssert = assert.eql(
+      testCase.method.toLowerCase(),
+      request.method.name.toLowerCase()
+    )
+
+    List(
+      bodyAssert,
+      headerAssert.pure[IO],
+      uriAssert.pure[IO],
+      methodAssert.pure[IO]
+    ).combineAll
+  }
+
+  private[compliancetests] def clientRequestTest[I, E, O, SE, SO](
+      endpoint: Endpoint[Op, I, E, O, SE, SO],
+      testCase: HttpRequestTestCase
+  ): ComplianceTest[IO] = {
+    type R[I_, E_, O_, SE_, SO_] = IO[O_]
+
+    val inputFromDocument = Document.Decoder.fromSchema(endpoint.input)
+    ComplianceTest[IO](
+      name = endpoint.id.toString + "(client|request): " + testCase.id,
+      run = {
+        val input = testCase.params
+          .map { inputFromDocument.decode(_).liftTo[IO] }
+          .getOrElse(IO.pure(().asInstanceOf[I]))
+
+        deferred[Request[IO]].flatMap { requestDeferred =>
+          val app = HttpRoutes
+            .of[IO] { case req =>
+              req.body.compile.toVector
+                .map(fs2.Stream.emits(_))
+                .map(req.withBodyStream(_))
+                .flatMap(requestDeferred.complete(_))
+                .as(Response[IO]())
+            }
+            .orNotFound
+
+          getClient(app).use { client =>
+            // avoid blocking the test forever...
+            val recordedRequest = requestDeferred.get.timeout(1.second)
```

Comment:
you may want to add the timeout to to the whole computation, instead of the request recording, because if the `endpoint.wrap(in)` was to never return, you'd be deadlocked. 

Unless you do `parProductR` (assuming it exists)